certbot-nginx/certbot_nginx/nginxparser.py

```diff
@@ -115,24 +131,120 @@ def load(_file):
     :rtype: NoneType
 
     """
-    return _file.write(dumps(blocks, indentation))
+    return _file.write(dumps(blocks))
+
+
+spacey = lambda x: (isinstance(x, str) and x.isspace()) or x == ''
+
+class UnspacedList(list):
+    """Wrap a list [of lists], making any whitespace entries magically invisible"""
+
+    def __init__(self, list_source):
+        # ensure our argument is not a generator, and duplicate any sublists
+        self.spaced = copy.deepcopy(list(list_source))
+
+        # Turn self into a version of the source list that has spaces removed
+        # and all sub-lists also UnspacedList()ed
+        list.__init__(self, list_source)
+        for i, entry in reversed(list(enumerate(self))):
+            if isinstance(entry, list):
+                sublist = UnspacedList(entry)
+                list.__setitem__(self, i, sublist)
+                self.spaced[i] = sublist.spaced
+            elif spacey(entry):
+                # don't delete comments
+                if "#" not in self[:i]:
+                    list.__delitem__(self, i)
+
+    def _coerce(self, inbound):
+        """
+        Coerce some inbound object to be appropriately usable in this object
+
+        :param inbound: string or None or list or UnspacedList
+        :returns: (coerced UnspacedList or string or None, spaced equivalent)
+        :rtype: tuple
+
+        """
+        if not isinstance(inbound, list):                      # str or None
+            return (inbound, inbound)
+        else:
+            if not hasattr(inbound, "spaced"):
+                inbound = UnspacedList(inbound)
+            return (inbound, inbound.spaced)
+
+
+    def insert(self, i, x):
+        item, spaced_item = self._coerce(x)
+        self.spaced.insert(self._spaced_position(i), spaced_item)
+        list.insert(self, i, item)
+
+    def append(self, x):
+        item, spaced_item = self._coerce(x)
+        self.spaced.append(spaced_item)
+        list.append(self, item)
+
+    def extend(self, x):
+        item, spaced_item = self._coerce(x)
+        self.spaced.extend(spaced_item)
+        list.extend(self, item)
+
+    def __add__(self, other):
+        l = copy.deepcopy(self)
+        l.extend(other)
+        return l
+
+    def __setslice__(self, i, j, newslice):
+        raise NotImplementedError("Slice operations on UnspacedLists not yet implemented")
+
+    def __setitem__(self, i, value):
+        if isinstance(i, slice):
+            raise NotImplementedError("Slice operations on UnspacedLists not yet implemented")
+        item, spaced_item = self._coerce(value)
+        self.spaced.__setitem__(self._spaced_position(i), spaced_item)
+        list.__setitem__(self, i, item)
+
+    def __delitem__(self, i):
+        self.spaced.__delitem__(self._spaced_position(i))
+        list.__delitem__(self, i)
+
+    def __deepcopy__(self, unused_memo):
+        l = UnspacedList(self[:])
+        l.spaced = copy.deepcopy(self.spaced)
+        return l
+
```

Comment:
I looked at the other methods defined by Python list objects and several are missing and will fall back to the list implementation. For example, the lack of `__mul__()` means that `UnspacedList(["foo", " ", "bar"])*3` is just the list `["foo", "bar", "foo", "bar", "foo", "bar"]` without a `.spaced` attribute.

In some cases, this is desirable; for example, for `__getitem__`, the behavior of the underlying list is exactly correct. It is arguably also correct for `__eq__` and other comparisons. In other cases, it may result in some confusion.

I see these list methods as not specifically implemented here: `__class__`, `__contains__`, `__delattr__`, `__delslice__`, `__doc__`, `__eq__`, `__format__`, `__ge__`, `__getattribute__`, `__getitem__`, `__getslice__`, `__gt__`, `__hash__`, `__iadd__`, `__imul__`, `__init__`, `__iter__`, `__le__`, `__len__`, `__lt__`, `__mul__`, `__ne__`, `__new__`, `__reduce__`, `__reduce_ex__`, `__repr__`, `__reversed__`, `__rmul__`, `__setattr__`, `__sizeof__`, `__str__`, `__subclasshook__`, `count`, `index`, `pop`, `remove`, `reverse`, `sort`.

Many of these are arguably right to use the underlying list implementation (for example, things like `index`, `count`, and `__len__` are credible). I am still concerned about at least `pop`, `remove`, `reverse`, and `sort`, which all modify the list. I tested trying to `reverse` an `UnspacedList`, and it caused the `.spaced` version to become severely desynchronized with the user-visible content.

I think the right answer is probably "anything that generates a new list with the entire content of the old list, or that modifies the existing object, should either be reimplemented or raise a `NotImplementedError`". So I have 4 specific examples of things I think need one or the other treatment. I think this is prudent even if the existing code doesn't call any of those methods, to avoid hard-to-debug cases if the parser is modified in the future.